boring/src/ssl/mod.rs

```diff
@@ -2975,6 +2994,27 @@ impl SslRef {
         }
     }
 
+    /// Sets a custom certificate store for verifying peer certificates.
+    ///
+    /// This corresponds to [`SSL_CTX_set0_verify_cert_store`].
+    ///
+    /// [`SSL_set0_verify_cert_store`]: https://docs.openssl.org/1.0.2/man3/SSL_CTX_set1_verify_cert_store/
+    pub fn set_verify_cert_store(&mut self, cert_store: X509Store) -> Result<(), ErrorStack> {
+        #[cfg(feature = "rpk")]
+        assert!(
+            !self.ssl_context().is_rpk(),
+            "This API is not supported for RPK"
+        );
+
+        unsafe {
+            let ptr = cert_store.as_ptr();
+            cvt(ffi::SSL_set0_verify_cert_store(self.as_ptr(), ptr) as c_int)?;
+            mem::forget(cert_store);
```

Comment:
I took a brief look at the manpage and it's not clear to me why we need to pass cert_store by value and then forget it, rather than just passing a reference. Is it just that this only happens once on startup and it's okay to leak? It might be nice to have a [// SAFETY: block](https://std-dev-guide.rust-lang.org/policy/safety-comments.html)

Reply:
Also for some reason when I inspect the first commit individually it says the function as a whole is new, but when I look at master it's already there and looking at the second commit it only shows the actual cert_store.as_ptr() change. weird.

Reply:
On master you may be seeing the `SSL_CTX_set0_verify_cert_store` variant

Reply:
There are generally two variants for each API - one that operates on `SSL_CTX` (applies to the server generally), and another that operates on `SSL` (connection specific)

Reply:
Re: calling mem::forget manually, it's a good question, but I'm mainly following prior art: https://github.com/sfackler/rust-openssl/blob/master/openssl/src/ssl/mod.rs#L3408-L3414

Reply:
It's potentially something to do with the destructor being set to `X509_STORE_free`, in this case `mem::forget` will actually free memory without running the destructor. I haven't dug in fully, but I assume we don't want to run `X509_STORE_free` here

Reply:
The more I look into it the more I believe that to be the case, `SSL_set0_verify_cert_store` bumps the refcount on the underlying pointer and boringssl becomes the responsible party for managing that memory. If we let the user program `drop` the owned type, it will actually run `X509_STORE_free` which will cause issues. By calling `mem::forget`, we eliminate the possibility of the user messing with the `X509Store` object after instantiating it on boringssl state

Reply:
Agree on not wanting to free here due to the contract of verify_cert_store specified in the manpage, but it seems like the X509Store [has a pointer variant](https://docs.rs/boring/latest/boring/x509/store/struct.X509StoreRef.html) that we could use and then leave the decision of leaking vs keeping around in a context/state/whatever struct for reuse up to the caller.

But I do not have as much experience as sfackler so there's a good chance they had a reason for doing it this way.

Reply:
It does indeed have a pointer variant, but the `X509StoreBuilder` needs to yield an owned `X509Store` back to the caller before it's actually registered on boringssl state. If it yielded a reference, nothing would actually be keeping it alive under the hood. I think that's why all of the APIs that actually _register_ the store on the underlying boringssl state take ownership of the `X509Store` and call `mem::forget` on it

Reply:
To summarize: boringssl takes ownership of the `X509Store` (as Rushil mentioned above) and gets freed by boringssl. Thanks for taking the time to help me understand!

Reply:
For posterity, this works by virtue of refcounted objects in c++. When initialized, a `X509Store` will have a refcount of 1. Any time the destructor is called, the refcount will be decremented and checked against `== 0` before the object is fully freed. By using the `set0` variant of the API, we "hand off" the object at refcount == 1 to c++ and leave the memory's lifetime at mercy of the next destructor call

Reply:
Given that this is a common pattern I still think it would be nice to document somewhere

Reply:
The intent should be conveyed by using [`<X509Store as ForeignType>::into_ptr`](https://docs.rs/foreign-types-shared/0.3.1/foreign_types_shared/trait.ForeignType.html#method.into_ptr)

Reply:
Opened https://github.com/cloudflare/boring/pull/267